runtime/lua/vim/lsp/types/protocol.lua

```diff
@@ -0,0 +1,4401 @@
+---folding ranges to display custom labels instead of the default text.
+---
+---@since 3.17.0
+---@field collapsedText? lsp.boolean
+
+---@class anonym38
+---The tags supported by the client.
+---@field valueSet lsp.DiagnosticTag[]
+
+---@class anonym40
+
+---@class anonym41
+---The client will send the `textDocument/semanticTokens/full/delta` request if
+---the server provides a corresponding handler.
+---@field delta? lsp.boolean
+
+---@class anonym39
+---The client will send the `textDocument/semanticTokens/range` request if
+---the server provides a corresponding handler.
+---@field range? lsp.boolean|anonym40
+---The client will send the `textDocument/semanticTokens/full` request if
+---the server provides a corresponding handler.
+---@field full? lsp.boolean|anonym41
+
+---@class anonym42
+---The properties that a client can resolve lazily.
+---@field properties lsp.string[]
+
+---@class anonym43
+---Whether the client supports additional attributes which
+---are preserved and send back to the server in the
+---request's response.
+---@field additionalPropertiesSupport? lsp.boolean
+
+---@class anonym44
+---@field range lsp.Range
+---@field placeholder lsp.string
+
+---@class anonym45
+---@field defaultBehavior lsp.boolean
+
+---@class anonym46
+---The range of the document that changed.
+---@field range lsp.Range
+---The optional length of the range that got replaced.
+---
+---@deprecated use range instead.
+---@field rangeLength? lsp.uinteger
+---The new text for the provided range.
+---@field text lsp.string
+
+---@class anonym47
+---The new text of the whole document.
+---@field text lsp.string
+
+---@class anonym48
+---@field language lsp.string
+---@field value lsp.string
+
+---@class anonym49
+---A language id, like `typescript`.
+---@field language lsp.string
+---A Uri {@link Uri.scheme scheme}, like `file` or `untitled`.
+---@field scheme? lsp.string
+---A glob pattern, like `*.{ts,js}`.
+---@field pattern? lsp.string
+
+---@class anonym50
+---A language id, like `typescript`.
+---@field language? lsp.string
+---A Uri {@link Uri.scheme scheme}, like `file` or `untitled`.
+---@field scheme lsp.string
+---A glob pattern, like `*.{ts,js}`.
+---@field pattern? lsp.string
+
+---@class anonym51
+---A language id, like `typescript`.
+---@field language? lsp.string
+---A Uri {@link Uri.scheme scheme}, like `file` or `untitled`.
+---@field scheme? lsp.string
+---A glob pattern, like `*.{ts,js}`.
+---@field pattern lsp.string
+
+---@class anonym52
+---The type of the enclosing notebook.
+---@field notebookType lsp.string
+---A Uri {@link Uri.scheme scheme}, like `file` or `untitled`.
+---@field scheme? lsp.string
+---A glob pattern.
+---@field pattern? lsp.string
+
+---@class anonym53
+---The type of the enclosing notebook.
+---@field notebookType? lsp.string
+---A Uri {@link Uri.scheme scheme}, like `file` or `untitled`.
+---@field scheme lsp.string
+---A glob pattern.
+---@field pattern? lsp.string
+
+---@class anonym54
```

Comment:
What's with all the `anonym` cases? I don't think we want those.

Reply:
those are for literal tables or however you wanna call them
e.g. here
![image](https://github.com/neovim/neovim/assets/81827001/ac30ec66-75fd-41b5-99eb-31d87585a7bb)

Reply:
I also would prefer `{name: string, version?: string}` but there you can't add documentation afaik
I opened an issue regarding this in the lua-ls repo

Reply:
The ones with the property name seems like the best approach fornow. That's exactly what I'm doing in neoconf to generate the types of all the different lsp server settings

Reply:
I just chose the `lsp.***.***` style like `---@class LSP.CompletionList.itemDefaults` in my Lua utility.

The benefit of this style is that can be used the partial types on userland.

If we choose the `Anonym` style, I think `lsp.Anonym` is better.

Reply:
I agree with @hrsh7th here. It's the same I use in neodev and neoconf.